bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java

```diff
@@ -91,87 +91,166 @@
+     *
+     * @param templateRegistry templateRegistry service.
+     */
+    protected void unsetTemplateRegistry(TemplateRegistry templateRegistry) {
+        this.templateRegistry = null;
+    }
+
+    protected void setStorageService(StorageService storageService) {
+        // this.storageService = storageService;
+        setDisabledRuleStorage(
+                storageService.<Boolean> getStorage("automation_rules_disabled", this.getClass().getClassLoader()));
+        setManagedProvider(new ManagedRuleProvider(storageService));
     }
 
     /**
-     * This method is used to register all {@link Rule}s provided via the {@link RuleProvider}, into the
-     * {@link RuleEngine}.
+     * Unbind the {@link StorageService} - called from DS.
      *
-     * @param provider a provider of {@link Rule}s.
-     * @throws RuntimeException
-     *             when passed module has a required configuration property and it is not specified in rule definition
-     *             nor
-     *             in the module's module type definition.
-     * @throws IllegalArgumentException
-     *             when a module id contains dot or when the rule with the same UID already exists.
+     * @param storageService
+     */
+    protected void unsetStorageService(StorageService storageService) {
+        // this.storageService = null;
+        removeManagedProvider(managedProvider);
+    }
+
+    /**
+     *
+     * @param templateProvider
      */
+    protected void addTemplateProvider(TemplateProvider templateProvider) {
+        templateUpdated(templateProvider.getTemplates(null));
+    }
+
+    protected void updateTemplateProvider(TemplateProvider templateProvider) {
+        templateUpdated(templateProvider.getTemplates(null));
+    }
+
+    @Override
+    protected void setEventPublisher(EventPublisher eventPublisher) {
+        super.setEventPublisher(eventPublisher);
+    }
+
+    @Override
+    protected void unsetEventPublisher(EventPublisher eventPublisher) {
+        super.unsetEventPublisher(eventPublisher);
+    }
+
     @Override
     protected void addProvider(Provider<Rule> provider) {
-        logger.info("Rule provider: {} is added.", provider);
+        LOGGER.info("Rule provider: {} is added.", provider);
         super.addProvider(provider);
+        if (provider instanceof ManagedRuleProvider) {
+            ruleRegistryReg = bundleContext.registerService(RuleRegistry.class.getName(), this, null);
+        }
     }
 
     @Override
-    protected void setManagedProvider(ManagedProvider<Rule, String> provider) {
-        super.setManagedProvider(provider);
-        logger.info("Rule Managed Provider: {} is added.", provider);
+    protected void removeProvider(Provider<Rule> provider) {
+        LOGGER.info("Rule provider: {} is removed.", provider);
+        if (provider instanceof ManagedRuleProvider) {
+            if (ruleRegistryReg != null) {
+                ruleRegistryReg.unregister();
```

Comment:
The RuleRegistry must not be registered if the managed provider is not available. Otherwise a rule from other provider with same uid as the ones from managed provider can be loaded into rule engine and all changes stored in the managed provider will be not visible (the rule from managed provider will not be available at all). For this reason the RuleEngine first has to be  initialized with rules from managed provider and after then it is registered and give change to other providers to make make add their rules in the RuleEngine.
May be the RuleRegistry must not be unregistered when managed provider is disappeared, because it can return available rules, only update command will throw exception